### PR TITLE
[NT-1371, NT-1331, NT-1391, NT-1400, NT-1436] Create/Update Backing with Add-Ons

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Refactors `PledgeViewModel` and all related code to support backing reward with add-ons. This PR started out with the purpose of adding navigation to the continue button when add-ons were selected and it quickly grew to include all of this due to the regressions that would have been created had this code not been refactored 🤷‍♂️. Apologies for the size of this PR.

# 🤔 Why

This allows backers to back rewards with add-ons and also to make use of the manage pledge flows for the backing that is created.

**Note:** Currently you can't edit your add-on selection, that will be in an upcoming PR.

# 🛠 How

- Refactored `PledgeViewModel` and attempted to simplify the way that totals are split.
- Added `PledgeShippingSummaryView`.
- Fixed all `PledgeViewModelTests`. I think that the values make more sense now.
- Added bonus support values to tracking properties.
- Updated `ManagePledgeViewModel` and tests.
- Added and updated snapshot tests for some views.

# 👀 See

| Before 🐛 | After 🦋 |
| --- | --- |
| image | image |

# ✅ Acceptance criteria

The checkout flow will need full regression testing but areas to focus on:

- [x] Pledge to No Reward, confirm all values.
- [ ] Pledge to a regular reward without shipping, confirm all values.
- [x] Pledge to a regular reward with shipping, confirm all values.
- [ ] Pledge to a reward with add-ons without shipping, confirm all values.
- [x] Pledge to a reward with add-ons with shipping, confirm all values.
- [x] Attempt all available update flows (update pledge, shipping, payment method) on all of the above, confirm values as you go.
- [x] Try all or some of the above using Apple Pay to checkout and also confirm that the values displayed in the sheet are correct.
- [x] Have fun.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift [hunk 1]
```diff
@@ -85,34 +104,80 @@ public final class RewardAddOnSelectionViewModel: RewardAddOnSelectionViewModelT
...
       .negate()
 
-    let totalSelectedQuantity = latestSelectedQuantities.map { quantities in
-      quantities.reduce(0) { accum, keyValue in
-        let (_, value) = keyValue
-        return accum + value
-      }
+    let totalSelectedAddOnsQuantity = Signal.combineLatest(
+      latestSelectedQuantities,
+      baseReward
```
Comment (jgsamudio) [inline Library/ViewModels/RewardAddOnSelectionViewModel.swift:112]: Is there a need to have the base reward here if it's filtered out after?

File: Library/ViewModels/RewardsCollectionViewModel.swift [hunk 2]
```diff
@@ -96,21 +94,24 @@ public final class RewardsCollectionViewModel: RewardsCollectionViewModelType,
...
-    .map { project, reward, refTag in
-      PledgeData(project: project, reward: reward, refTag: refTag)
+    .filter { project, reward, _ in
+      project.state == .live && !userIsBacking(reward: reward, inProject: project)
+    }
+    .map { project, reward, refTag -> (PledgeViewData, Bool) in
+      let data = PledgeViewData(
+        project: project,
+        rewards: [reward],
+        selectedQuantities: [reward.id: 1],
```
Comment (jgsamudio) [inline Library/ViewModels/RewardsCollectionViewModel.swift:104]: Is this going to change in the future? Should we be setting this to 1?

File: Library/ViewModels/RewardAddOnSelectionViewModel.swift (hunk 1 above)
Comment (justinswart) [reply Library/ViewModels/RewardAddOnSelectionViewModel.swift:112]: Yeah, we need it to know its ID in order to filter it out.

File: Library/ViewModels/RewardsCollectionViewModel.swift (hunk 2 above)
Comment (justinswart) [reply Library/ViewModels/RewardsCollectionViewModel.swift:104]: Well-spotted. This is just a somewhat contrived way of abstracting the behaviour for regular rewards and add-ons based rewards. We know (for now) that the base reward will always be selected once.

**Pull request merged**